### ISSUE TITLE
abstract_ui vs error messages

### ISSUE THREAD

**Issue by coolwanglu** [neovim/neovim#1984]
Sometimes nvim would show some error messages before the main windows shows up, e.g. `Warning: cannot find word list ...`, and expect the Enter key to be pressed before nvim can proceed.

However there are no events for remote UI for such messages, so the UI window cannot be initialized until the user press a key, which is confusing.

Shall we add one more redraw event? Or just print those messages to stderr for abstract_ui?

**Comment by tarruda** [comment]
> However there are no events for remote UI for such messages, so the UI window cannot be initialized until the user press a key, which is confusing.

:+1:

The problem is that the messages are printed before the UI is attached, but there's currently no way to delay the startup process(and thus hold the messages until the embedder's attach its own UI)

I will fix this problem after #1820 is merged, which should happen between sunday and monday.

**Comment by justinmk** [comment]
@tarruda While working on #1802 I noticed that Vim collects error messages for the GUI and defers them until the GUI is ready. I (temporarily?) removed that code here:

https://github.com/justinmk/neovim/commit/882df042e019177ddaec91c1ef21eb13ed4d70b7

Maybe that code can be used for this issue, unless I misunderstand.

**Comment by tarruda** [comment]
@justinmk its not ok to remove `mch_errmsg` as it is needed for the new UI architecture(Especially after #1820, which puts TUIs and GUIs on the same level).

It probably needs a better name though

**Comment by justinmk** [comment]
> Sometimes nvim would show some error messages before the main windows shows up, e.g. Warning: cannot find word list ..., and expect the Enter key to be pressed before nvim can proceed.

@equalsraf How does nvim-qt deal with this? Is it still a problem?

**Comment by equalsraf** [comment]
> @equalsraf How does nvim-qt deal with this?

it doesn't, there is no special handling for this issue, other than a fatal error if we cannot get a response from Neovim in a while, but see below.

> Is it still a problem?

Previously what happened was that if nvim blocked waiting for input on an early error them the GUI would never be able to connect, because nvim was blocked.

This is still an issue, it definitely happens with pynvim. Strangely it doesn't seem to happen with nvim-qt, e.g.

    nvim-qt -- --cmd "invalid vim cmd"

The GUI starts fine. The error message is only visible when calling :messages.

If you call the TUI you get

    $ export NVIM_LISTEN_ADDRESS=nvimsock
    $ nvim --cmd "invalid vim cmd"
    Error detected while processing pre-vimrc command line:
    E492: Not an editor command: invalid vim cmd
    Press ENTER or type command to continue

And the TUI blocks waiting for input. If you then try to connect to the running instance

    $ nvim-qt --server nvimsock

the GUI connects, the TUI stops waiting for input and everything starts up i.e. nvim starts in normal mode. The behaviour is identical with init.vim errors.

It turns out this **is a bug in nvim-qt**. When it starts up it makes an unintended call to vim_input("\<FocusGained\>") BEFORE attachment, this ends up unblocking the prompt. A call to vim_input("\n") also works as a workaround.

**Comment by teto** [comment]
I might be willing to take on this. 
I just would like to understand the approach here: rather then reestablishing `mch_errmsg` and stores messages until an UI connects, could it make sense to generalize the approach to all UI events ?
What about adding some kind of a fake sink UI that just stores messages (up to a limit/filters redundant) and forwards them as soon as a real UI connects ? It would allow to defer the connection of the UI but I don't know if it is of interest.

**Comment by justinmk** [comment]
> What about adding some kind of a fake sink UI that just stores messages (up to a limit/filters redundant) and forwards them as soon as a real UI connects ?

How's that different than "reestablishing mch_errmsg and stores messages until an UI connects"? I'm ok with either approach but the "fake UI sink" sounds a bit more complicated without a benefit AFAICT.

**Comment by teto** [comment]
good, was just wondering if there was a scenario where it could be interesting. I'll stick to the simple version then.

**Comment by justinmk** [comment]
Duplicate of #3901 .

This is mostly addressed by https://github.com/neovim/neovim/pull/9024 though it might still be a good idea to "collect" messages during startup.
